component/loki/source/api/api.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/weaveworks/common/server"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.api",
+		Args: Arguments{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+type Arguments struct {
+	HTTPAddress          string              `river:"http_address,attr"`
+	HTTPPort             int                 `river:"http_port,attr"`
+	ForwardTo            []loki.LogsReceiver `river:"forward_to,attr"`
+	Labels               map[string]string   `river:"labels,attr,optional"`
+	RelabelRules         relabel.Rules       `river:"relabel_rules,attr,optional"`
+	UseIncomingTimestamp bool                `river:"use_incoming_timestamp,attr,optional"`
+	// TODO: allow to configure other Server fields in a dedicated block, to match promtail's
+	//       https://grafana.com/docs/loki/next/clients/promtail/configuration/#server
+}
+
+func (a *Arguments) labelSet() model.LabelSet {
+	labelSet := make(model.LabelSet, len(a.Labels))
+	for k, v := range a.Labels {
+		labelSet[model.LabelName(k)] = model.LabelValue(v)
+	}
+	return labelSet
+}
+
+type Component struct {
+	opts         component.Options
+	entriesChan  chan loki.Entry
+	unregisterer *util.Unregisterer
+
+	rwMut      sync.RWMutex
+	args       Arguments
+	pushTarget *lokipush.PushTarget
+
+	// Use separate receivers mutex to address potential deadlock when Update drains the current server.
+	// e.g. https://github.com/grafana/agent/issues/3391
+	receiversMut sync.RWMutex
+	receivers    []loki.LogsReceiver
+}
+
+func New(opts component.Options, args Arguments) (component.Component, error) {
+	c := &Component{
+		opts:         opts,
+		args:         args,
+		entriesChan:  make(chan loki.Entry),
+		receivers:    args.ForwardTo,
+		unregisterer: util.WrapWithUnregisterer(opts.Registerer),
+	}
+	err := c.Update(args)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) (err error) {
+	defer func() {
+		err = c.stop()
+	}()
+
+	for {
+		select {
+		case entry := <-c.entriesChan:
+			c.receiversMut.RLock()
+			receivers := c.receivers
+			c.receiversMut.RUnlock()
+
+			for _, receiver := range receivers {
+				select {
+				case receiver <- entry:
+				case <-ctx.Done():
+					return
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	newArgs, ok := args.(Arguments)
+	if !ok {
+		return fmt.Errorf("invalid type of arguments: %T", args)
+	}
+
+	c.receiversMut.Lock()
+	c.receivers = newArgs.ForwardTo
+	c.receiversMut.Unlock()
+
+	newPushTargetConfig := c.pushTargetConfigForArgs(newArgs)
+
+	c.rwMut.Lock()
+	defer c.rwMut.Unlock()
+
+	pushTargetNeedsUpdate := c.pushTarget == nil || !reflect.DeepEqual(c.pushTarget.CurrentConfig(), *newPushTargetConfig)
+	if !pushTargetNeedsUpdate {
+		c.args = newArgs
+		return nil
+	}
```

Comment:
I think the address here will always be different; do we want to compare the struct itself?

And if no changes are needed, then we don't need to update c.args either, right?

Reply:
My point was that reflect.DeepEqual would compare the addresses and the second part of the clause would _always_ turn to true, right? Shouldn't we have reflect.DeepEqual called on the dereferenced values?

Reply:
I did wonder the same thing, but `reflect.DeepEqual` will follow the pointers and compare the actual values they are pointing to. I have a test for this here: https://github.com/grafana/agent/blob/21fcc26d7e86118c14c091b9921e7f4d72cf9552/component/loki/source/api/api_test.go#L230

I also have a simple example: https://go.dev/play/p/yBrKf2pyms9.

I think it goes via this code path: https://github.com/golang/go/blob/master/src/reflect/deepequal.go#L127

Reply:
Ok, thanks for the heads up!